rfc(0011): mapping events to alerts

ncphillips opened the pull request:
Open for review until August 28th

chrisdmacrae commented:
Two questions:

- how does the example plugin bind its default alerts to the CMS
- will an API for adding alerts dynamically during runtime still exist?

ncphillips commented:
> * how does the example plugin bind its default alerts to the CMS

The CMS does that [when the API is registered](https://github.com/tinacms/rfcs/pull/12/files#diff-5281be4515b982e35e2e6931d7154e09R124-R134)

> * will an API for adding alerts dynamically during runtime still exist?

Yes this is a backwards compatible change.

dwalkr commented:
> These Alerts can be replaced at runtime

The RFC doesn't appear to indicate how this would be done. Would a user need to re-instantiate the CMS?

I'm thinking about a multilingual TinaCMS setup, hypothetically using either (1) the browser's localization preferences, or (2) a toolbar widget / setting to switch the UI language

dwalkr commented on 0011-mapping-events-to-alerts.md:
```diff
@@ -0,0 +1,145 @@
+
+### Default Alert Maps in APIs
+
+Any API registered with the CMS could define default alerts for it's events.
```
```suggestion
Any API registered with the CMS could define default alerts for its events.
```
so sorry

dwalkr commented:
> But the proposed API does not expose that mapping

Exactly! I think this RFC should include a `replaceAlerts` method or something along those lines

ncphillips commented:
Updated the implementation idea and added a "Updating dynamically" section. This stuff will live on the `Alerts` class not the `CMS` class. In hindsight that's obviously a better spot.

dwalkr commented:
I'm going to leave some comments/questions about implementation. Not trying to get too deep in the weeds here as we can worry about specific implementation later, but I think it's relevant to some degree

dwalkr commented on 0011-mapping-events-to-alerts.md:
```diff
@@ -0,0 +1,162 @@
+    // ...
+
+    if (this.api.alerts) {
+      this._eventsToAlerts = {
```
this could just call `this.alerts.setMap` now

dwalkr commented on 0011-mapping-events-to-alerts.md:
```diff
@@ -0,0 +1,162 @@
+};
+
+class TinaCMS extends CMS {
+  private _eventsToAlerts: EventsToAlerts = {};
```
`_eventsToAlerts`  wouldn't exist on `TinaCMS` now but be moved to the `Alerts` class, right?

dwalkr commented on 0011-mapping-events-to-alerts.md:
```diff
@@ -0,0 +1,162 @@
+}
+
+class Alerts {
+  constructor(private events: EventBus, alerts = {}) {
```
I assume the constructor would call `this.setMap(alerts)`

dwalkr commented on 0011-mapping-events-to-alerts.md:
```diff
@@ -0,0 +1,162 @@
+      if (toAlert) {
+        const { level, message, timeout } = toAlert(event);
+
+        this.alerts.add(level, message, timeout);
```
```suggestion
        this.add(level, message, timeout);
```

ncphillips commented:
Thank you! That was sloppy fixing them up now

chrisdmacrae commented:
Based on the current iteration, this will work perfectly with the experimental i18n plugin, which is an indication to me it’s pretty close.
